Remove LocalizeStaticMixin (#3499)
BrightspaceUI/core: mixins/localize/README.md

Author: This might be best viewed as a side-by-side diff. Mostly I just removed the static-specific stuff, but also made the language a bit more formal by removing "your" and "you". There's now a dedicated OSLO section, and I removed some of the more obscure sections down to an "Advanced".

# LocalizeMixin

The `LocalizeMixin` allows text in components to be displayed in the user's preferred language.

## Language Resources

Resources are stored as key-value JSON objects.

### Keys

The key should succinctly and uniquely describe the text being localized. `camelCase` is recommended, although `snake_case` and `kebab-case` are also supported.

For large projects, terms may be grouped using the `:` character. For example: `parentGroup:subGroup:termName`.

### Values

Term values must conform to the [ICU Message Syntax](https://formatjs.io/docs/core-concepts/icu-syntax/) format. It supports features such as: [simple arguments](https://formatjs.io/docs/core-concepts/icu-syntax/#simple-argument), the [`{select}` format](https://formatjs.io/docs/core-concepts/icu-syntax/#select-format) and [pluralization](https://formatjs.io/docs/core-concepts/icu-syntax/#plural-format).

> **Note:** Avoid using the ICU Message Syntax number, date and time formatting functionality. Brightspace allows customization of how these are localized, so use [@brightspace-ui/intl](https://github.com/BrightspaceUI/intl)'s `formatNumber`, `formatDate` and `formatTime` instead.

### Files

> Store localization resources in their own directory with nothing else in it. There should be one JavaScript file for each language.

Reviewer: ```suggestion
Store localization resources in their own directory with nothing else in it. There should be one JavaScript file for each supported locale.
```

};
```
```javascript
// fr.js
export default {
  "hello": "Bonjour {firstName}!"
};
```

Always provide language resources for base languages (e.g. `en`, `fr`, `pt`). That way, if the user prefers a regional language (e.g. `pt-br`) that isn't recognized, it can fall back to the base language (`pt`).

## Using `LocalizeMixin`

The component should import and extend `LocalizeMixin`:

```javascript
import { LocalizeMixin } from '@brightspace-ui/core/mixins/localize/localize-mixin.js';

class MyComponent extends LocalizeMixin(LitElement) {

}
```

Implement a static getter for `localizeConfig` that defines `importFunc()`. It will be passed a language which can subsequently be dynamically imported from the location of the component's resources.

The dynamic import path must be relative.

```javascript
static get localizeConfig() {
  return {
    importFunc: async lang => (await import(`../lang/${lang}.js`)).default,
  };
}
```

### `localize()`

The `localize()` method is used to localize a piece of text in the component's `render()` method.

If the localized string contains arguments, pass them as a key-value object as the 2nd parameter:

```javascript
render() {

>   const message = this.localize('hello', { firstName: 'Mary' };
>   return html`<p>${message)}</p>`;

Reviewer 2: ```suggestion
  const message = this.localize('hello', { firstName: 'Mary' });
  return html`<p>${message}</p>`;
```

Similarly `<em>` *emphasizes* a particular piece of text (browsers show this visually using italics), whereas `<i>` only italicizes the text visually without emphasis.

> ```json
> {
>   "myTerm": "This is <b>bold</b> but <em>not</em> all that <strong>important</strong>."
> };

Reviewer: No semicolon in JSON:
```suggestion
```json
{
  "myTerm": "This is <b>bold</b> but <em>not</em> all that <strong>important</strong>."
}
```

But maybe better as:
```suggestion
```javascript
{
  myTerm: "This is <b>bold</b> but <em>not</em> all that <strong>important</strong>."
};
```

> ```

Author: @svanherk specific example based on your feedback in demos yesterday

Reviewer 2: Thanks! Does it make sense to also show that the usage is just `this.localizeHTML('myTerm')` with nothing passed in as the second parameter?  Maybe that's overkill

Reviewer 3: I think that could be helpful 👍.
